File: Project4/cpp/src/ising.cpp
#include <iostream>
#include <random>
#include <armadillo>
#include <array>
#include <random>
#include <string>
#include <chrono>
#include <thread>
#include <stdexcept>
#include "ising.h"
#include "metamodel.h"

int magnetization(const arma::imat& A){
  // Return the magnetization of a state described by A, which is simply the sum
  // of the spins/elements. This function is somewhat redundant.
  int M = 0;
  for(unsigned int i = 0; i < A.n_cols; i++){
    for(unsigned int j = 0; j < A.n_rows; j++){
      M += A(i,j);
    }
  }
  return M;
}

arma::imat setInitialStateRandom(unsigned int N){
  // Generate armadillo integer matrix with uniform random negative or positive
  // ones.
  std::mt19937 gen;
  gen.seed(1233);
  std::uniform_real_distribution<double> dist(0.0,1.0);

  arma::imat state = arma::zeros<arma::imat>(N, N);
  for(unsigned int row = 0; row < N; row++){
    for(unsigned int col = 0; col < N; col++){
      if(dist(gen) < 0.5){
        state(row,col) = -1;
      }
      else{
        state(row,col) = 1;
      }
    }
  }
  return state;
}

arma::imat setInitialStateOrdered(unsigned int N, int fill){
  arma::imat state = arma::ones<arma::imat>(N, N);
  state.fill(fill);
  return state;
}

arma::imat setInitialState(const Metamodel& model){
  std::string initOrient = model.initialOrientation;
  if(initOrient == "random"){
    return setInitialStateRandom(model.N);
  }
  else if(initOrient == "up"){
    return setInitialStateOrdered(model.N, 1);
  }
  else if(initOrient == "down"){
    return setInitialStateOrdered(model.N, -1);
  }
  else{
    throw std::invalid_argument("Invalid initial state input. Valid inputs are 'random', 'up' and 'down'.");
  }
}

unsigned int period(int x, unsigned int N){
  unsigned int a = ((x+N)%N);
  if(a == 0){
    return 0;
  }
  else{
    return a;
  }
}

int PeriodicBoundary(int i, int limit, int add){
  return (i+limit+add) % limit;
}

int sumNeighbors(unsigned int i, unsigned int j, const arma::imat& A){
  unsigned int N = A.n_cols;
  int s = 0;
  s += A(period(i-1, N), j)  // Left neighbor
    +  A(i, period(j+1, N))  // Lower neighbor
    +  A(i, period(j-1, N))  // Upper neighbor
    +  A(period(i+1, N), j); // Right neighbor
  return s;
}

int totalEnergy(const arma::imat& A){
  int E = 0;
  int N = A.n_cols;
  for(unsigned int i = 0; i < A.n_cols; i++){
    for(unsigned int j = 0; j < A.n_rows; j++){
      E -= A(i,j)*(A(PeriodicBoundary(i, N, -1)) +
                   A(PeriodicBoundary(j, N, -1)));
    }
  }
  return E;
}

std::map<int,double> makeProbabilities(double beta){
  std::array<int, 5> dEStates{-8, -4, 0, 4, 8};
  std::map<int,double> probabilities;
  for(const auto& dE: dEStates){
    probabilities[dE] = exp(-beta*dE);
  }
  return probabilities;
}

void ising(const Metamodel& model){
  int seed = model.seed;
  std::cout << "Temperature is " << model.temperature << std::endl;
  // Setup RNG generator
  std::mt19937 gen;
  gen.seed(seed);
  std::uniform_real_distribution<double> random(0.0, 1.0);

  // Setup and run simulation of the Ising model. Unpack some vals from model.
  unsigned int N = model.N;      // Lattice size (N x N)
  unsigned int M = model.M;         // Number of MC-iterations
  int m, n, dE;

  // Create probability map for possible energy changes
  const auto probabilities = makeProbabilities(model.beta);

  // Initialize arrays and initial values
  arma::imat state = setInitialState(model);
  std::vector<arma::imat> states (model.M);
  states[0] = state; // Saves every state

  std::vector<double> energies (model.M);         // Save energy for each MC Cycle
  double energy = totalEnergy(state);
  energies[0] = energy;

  std::vector<int> magmoments (model.M);          // Hold magnetization for each MC
  int magmoment = magnetization(state);
  magmoments[0] = magmoment;

  std::vector<int> nFlippedVec (model.M-1);
  int nFlipped;

  unsigned int nSpinsTot = N*N;
  // Loop through MC-cycles
  for(unsigned int i = 1; i < M; i++){
    nFlipped = 0;
    for(unsigned int j = 0; j < nSpinsTot; j++){
        // Choose row (m) and column (n) index to perturb/flip given acceptance
        m = static_cast<int>(random(gen)*N);
        n = static_cast<int>(random(gen)*N);
        // Calculate consequent change of energy
        dE = 2*state(m,n)*sumNeighbors(m,n,state);
        // Metropolis algorithm
        if(dE < 0 || random(gen) <= probabilities.at(dE)){
          state(m,n) *= -1;
          energy     += dE;
          magmoment  += 2*state(m,n);
          nFlipped   += 1;
        }
      }
      // Update states
      states[i]        = state;
      energies[i]      = energy;
      magmoments[i]    = fabs(magmoment);
      nFlippedVec[i-1] = nFlipped;
  }
  // Save all data to binary files
  model.save(states, energies, magmoments, nFlippedVec);
  // Write metadata to be read by Python before reading binary files
  model.write();
}

void isingParallel(std::vector<double>& expectationValues, const Metamodel& model){
    // Setup RNG generator
    std::random_device rd;
    std::mt19937_64 gen(rd());
    std::uniform_real_distribution<double> random(0.0,1.0);

    // Setup and run simulation of the Ising model. Unpack some vals from model.
    unsigned int N = model.N;      // Lattice size (N×N)
    unsigned int M = model.M;      // Number of MC-iterations
    int m, n, dE;

    // Create probability map for possible energy changes
    const auto probabilities = makeProbabilities(model.beta);

    // Initialize values
    arma::imat state = setInitialState(model);
    double energy    = totalEnergy(state);
    int magMoment    = magnetization(state);

    // Loop through MC-cycles
    const unsigned int numSpinsTot = N*N;
    for(unsigned int i = 1; i <= M; i++){
        // Loop over the number of nodes in the lattice
        for(unsigned int j = 0; j < numSpinsTot; j++){
            // Choose row (m) and column (n) index to perturb/flip given acceptance
            m = static_cast<int>(random(gen)*N);
            n = static_cast<int>(random(gen)*N);

            // Calculate consequent change of energy
            dE = 2*state(m,n)*sumNeighbors(m,n,state);

            // Metropolis algorithm
            if(dE < 0 ||random(gen) <= probabilities.at(dE)){
                state(m,n) *= -1;
                energy     += dE;
                magMoment  += 2*state(m,n);
            }
        }
        expectationValues[0] += energy;
        expectationValues[1] += energy*energy;
        expectationValues[2] += magMoment;
        expectationValues[3] += magMoment*magMoment;
        expectationValues[4] += fabs(magMoment);
    }
}

File: Project4/julia/src/ising.jl
module Ising
export neighbours, ising!, hamiltonian, Metadata

using  MPI
include("./metamodel.jl")
using .Metamodel: Metadata, writedata!

pprint(x) = show(IOContext(STDOUT, limit=false), "text/plain", x)
function pprintln(x)
    pprint(x)
    println("")
end

function mprint(x)
    for i in 1:size(x, 1)
        for j in 1:size(x, 2)
            if x[i, j] == 1
                print("██")
            else
                print("▒▒")
            end
        end
        print("\n")
    end
end


function neighbours(i, j, A)
    N = size(A, 1)
    function period(x)
        # First ensure that the number is not negative
        # then wrap it around, and add 1 since Julia
        # arrays start at 1
        a = (x + N+1) % (N+1)
        if a == 0
            1
        else
            a
        end
    end
    [A[period(i-1), j], A[i, period(j-1)], A[i, period(j+1)], A[period(i+1), j]]
end

function hamiltonian(A, J)
    n, m = size(A)
    E = 0
    for i ∈ 1:n, j ∈ 1:m
        E -= A[i, j]*sum(neighbours(i, j, A))
    end
    return J*E
end

function hamiltonian(A, J, h)
    hamiltonian(A, J) - h*sum(A)
end


function constructlattice(model)
    # Use Int8 to use as little memory as possible
    if model.initialorientation == :random
        # Construct the lattice with randomly oriented spins.
        rand(model.rng, convert(Vector{Int8}, [-1, 1]),
             (model.latticesize, model.latticesize))
    elseif model.initialorientation == :up
        ones(Int8, (model.latticesize, model.latticesize))
    elseif model.initialorientation == :down
        -ones(Int8, (model.latticesize, model.latticesize))
    else
        throw(ArgumentError("Initial orientation $(model.initialorientation) is not valid."))
    end
end

function initializesystem(model::Metadata)
    N = model.latticesize
    M = model.MCiterations
    J = model.J
    h = model.h
    temperature = model.temperature
    saveperiod = model.saveperiod
    β = 1.0/(1*temperature)

    Λ = constructlattice(model)

    # Since the state space and hence energy space is discrete,
    # the probabilities can be precalculated
    # This is the entire state space to ΔE
    ΔEstates = [-4J, -2J, 0, 2J, 4J]
    # And the probability for each ΔE
    ratio = Dict(i => exp(-β*i) for i in ΔEstates)

    # Matrices to store data for analysis
    if model.version == :parallel
        evolution = zeros(Int8, (M, N, N))
        energies = zeros(Float64, (M-1)*N^2+1)
        magneticmoment = zeros(Float64, (M-1)*N^2+1)
    elseif model.version == :serial
        evolution = zeros(Int8, (N, N))
        energy = 0.0
        magneticmoment = 0.0
    else
        throw(ArgumentError("Unknown model: $(model.version)"))
    end
    (lattice, ratios, evolution, energies, magneticmoment)
end


function ising!(model::Metadata, version::Symbol)
    if version == :parallel
        isingparallel!(model)
    elseif version == :serial
        isingserial!(model)
    else
        throw(ArgumentError("$version is not supported"))
    end
end


function isingserial!(model::Metadata)
    N = model.latticesize
    M = model.MCiterations
    Λ, ratio, evolution, energies, magneticmoment = initializesystem(model)

    # Initial values
    evolution[1, :, :] = Λ
    energies[1] = hamiltonian(Λ, J)
    magneticmoment[1] = sum(Λ)

    # Metropolis Algorithm
    @time for m in 2:M
        counter = 1
        for index in eachindex(Λ)
            n = (N*N)*(m-2) + counter + 1
            i, j = rand(model.rng, 1:N, (1, 2))
            ΔE = J*Λ[i, j]*sum(neighbours(i, j, Λ))
            if ΔE < 0 || rand(model.rng) <= ratio[ΔE]
                Λ[i, j] = -Λ[i, j]
                energies[n] = energies[n-1] + ΔE
                magneticmoment[n] = magneticmoment[n-1] + 2*Λ[i, j]
            else
                energies[n] = energies[n-1]
                magneticmoment[n] = magneticmoment[n-1]
            end
            counter = counter + 1
        end
        evolution[m, :, :] = Λ
    end
    writedata!(model, evolution, energies, magneticmoment)
end

function isingparallel!(model::Metadata)
    N = model.latticesize
    M = model.MCiterations
    Λ, ratio, evolution, energies, magneticmoment = initializesystem(model)

    # Initial values
    energies = hamiltonian(Λ, J)
    magneticmoment = sum(Λ)

    # Metropolis Algorithm
    @time for m in 2:M
        for index in eachindex(Λ)
            i, j = rand(model.rng, 1:N, (1, 2))
            ΔE = J*Λ[i, j]*sum(neighbours(i, j, Λ))
            if ΔE < 0 || rand(model.rng) <= ratio[ΔE]
                Λ[i, j] = -Λ[i, j]
                energy += ΔE
                magneticmoment += 2*Λ[i, j]
            end
        end
    end
    println("Energy $energy")
    println("Magnetic moment $magneticmoment")
    mprint(Λ)
    # writedata!(model, Λ, energy, magneticmoment)
end

end
